lib/index.d.ts

```diff
@@ -0,0 +1,99 @@
+type i53 = number;
+type i24 = number;
+
+type Ty = Array<i53>;
+
+type Column = Array<unknown>;
+
+type Archetype = {
+	id: number,
+	edges: {
+		[key: i53]: {
+			add: Archetype,
+			remove: Archetype,
+		},
+	},
+	types: Ty,
+	type: string | number,
+	entities: Array<number>,
+	columns: Array<Column>,
+	records: { [key: number]: number },
+}
+
+type ArchetypeMap = {
+	cache: Array<ArchetypeRecord>,
+	first: ArchetypeMap,
+	second: ArchetypeMap,
+	parent: ArchetypeMap,
+	size: number,
+}
+
+type ArchetypeRecord = number;
+
+export type Entity<T = unknown> = number & { __nominal_type_dont_use: T }
+
+type EntityIndex = {
+    dense: {
+        [key: i24]: i53
+    };
+    sparse: {
+        [key: i53]: Record
+    }
+}
+
+type Record = {
+	archetype: Archetype,
+	row: number,
+	dense: i24,
+	componentRecord: ArchetypeMap,
+}
+
+type ExtractFromLuaTuple<T> = T extends LuaTuple<infer U> ? U : never;
+
+type QueryShim<T extends any[]> = {

```

Comment:
It should extend with `IterableFunction<LuaTuple[Entity, ...Pack]>`, where Pack is the inferred generics to simplify this type. This will require changing the `query` as well.

Don't need to add `Shim` to the type name. 

It also doesn't matter too much what they put in `without` as long as they are just entities so it could just be `...components: Entity[]` since it will just return `self` with the same pack.

Reply:
This type appears to fit the requirements. I've put it in pending review.

```ts
type Query<T extends Entity[]> = {
    without: (...components: Entity[]) => Query<T>;
} & IterableFunction<LuaTuple<[Entity, ...T]>>
```